### ISSUE TITLE
Issue with running FIO with DirectIO on Solaris 11.3

### ISSUE THREAD

**Issue by rmathara** [axboe/fio#136]
Trying to run FIO for solaris (2.0.14 FIO version) on a solaris VM (solaris11.3) using solarisaio and 
direct=1 and getting below error :
fio: the file system does not seem to support direct IO
Trying to run against a raw device, below is my job file and running as :
..
root@a7ec2c:~/fio#  ./fio --output-format=json --output=test1.out small_lat.job
fio: the file system does not seem to support direct IO
...
cat small_lat.job :
..
[global]
ioengine=solarisaio
filename=/dev/rdsk/c2t1d0s2

[random_write_test]
stonewall
overwrite=1
rw=randwrite
blocksize=8k
direct=1
iodepth=512
runtime=180
time_based=1
..

Also, I couldn't locate a later version 2.2.10 and later for Solaris, appreciate any pointers on where I can get a later FIO version for Solaris.

Thanks

**Comment by rmathara** [comment]
Can I get some feedback on above please! .. Thanks

**Comment by kevinkuehnert** [comment]
ZFS doesn't support Direct IO so the issue you are seeing is a file system limitation and not an issue with FIO.

Hope that helps.

**Comment by rmathara** [comment]
Hi, I am trying to test against a raw device (/dev/rdsk/c2t1d0s2) which doesn't have any filesystem on the Luns. My Solaris OS disk on VM  is using ZFS bt my FIO tests are against the raw devices. 
Thanks

**Comment by sitsofe** [comment]
@rmathara : This bug report seems to stem from a misunderstanding - if you have access to a raw (uncached) device you don't need to set `direct=1` because the I/O isn't being cached by the OS.

On Solaris, fio attempts to use the `directio` ioctl when you set `direct=1`. `directio`  (https://docs.oracle.com/cd/E19683-01/816-0213/6m6ne37so/index.html ) is a **filesystem** request to make a particular file descriptor's accesses bypass OS caches. Without it, uncached access to file data would be limited to an all or nothing mount option which would affect all files and require remounting to change. When you get down to the device level, Solaris provides both cached (`/dev/dsk/`) and uncached (`/dev/rdsk/`) versions of disks. As the user/program can already pick whether they want accesses to be cached (by choosing the appropriate `/dev/` path) there's no need to allow the user to request uncached accesses on a cached `/dev/dsk/` device. Further, it now doesn't make sense to allow the user to turn off cached accesses on the already uncached `/dev/rdsk/` devices.

On Linux, block devices are cached and (in general) there aren't raw (uncached) versions, so it makes sense to allow a program to use O_DIRECT (which is what fio uses when you set `direct=1` on regular Linux ioengines) to ask for uncached I/O to the device (in addition to files within a filesystem when the filesystem supports it).

Given the above, is there an issue?

TL;DR: Using fio's `direct=1` **on Solaris** only makes sense when trying to access files within a filesystem (assuming the filesystem supports it). For disks, just use the `/dev/rdsk/` version and leave fio's `direct` option at 0.

**Comment by sitsofe** [comment]
Perhaps this has been addressed by the improved warning message added by commit https://github.com/axboe/fio/pull/269/commits/42f1ee68ceec87fbbfdc4972c35d3cdf7c08d9f6 when Solaris `directio` fails...

**Comment by sitsofe** [comment]
Closing due to lack of reply from reporter. If this issue is still happening with the latest fio (see https://github.com/axboe/fio/releases to find out which version that is) please reopen. Thanks!
